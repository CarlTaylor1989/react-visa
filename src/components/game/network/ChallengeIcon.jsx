import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import Tooltip from '../../generic/tooltip/Tooltip';
import {
  CHALLENGE_PERFECT_STATUS,
  CHALLENGE_COMPLETION_STATUS,
  CHALLENGE_UNAVAILABLE_STATUS,
  CHALLENGE_AVAILABLE_CLASS,
  CHALLENGE_COMPLETED_CLASS,
  CHALLENGE_PERFECT_CLASS
} from '../../../config/constants';

import './ChallengeIcon.scss';

export class ChallengeIcon extends Component {
  constructor(props) {
    super(props);
    this.state = {
      status: '',
      tooltipStatus: CHALLENGE_AVAILABLE_CLASS
    };
  }

  componentDidMount() {
    this.updateChallengeState();
  }

  componentDidUpdate(prevProps) {
    const { lastChallengesModified, challengeId, mapInitialised } = this.props;
    if ((prevProps.lastChallengesModified !== lastChallengesModified
      && lastChallengesModified.includes(challengeId))
      || (!prevProps.mapInitialised && mapInitialised)) {
      this.updateChallengeState();
    }
  }

  updateChallengeState() {
    const {
      challengeId,
      familyId,
      challenges
    } = this.props;
    const { status } = this.state;
    const found = challenges.find(challenge => challenge.familyId === familyId
      && challenge.id === challengeId);

    let newStatus = '';
    if (found && found.status > CHALLENGE_UNAVAILABLE_STATUS) {
      switch (found.status) {
        case CHALLENGE_PERFECT_STATUS:
          newStatus = CHALLENGE_PERFECT_CLASS;
          break;
        case CHALLENGE_COMPLETION_STATUS:
          newStatus = CHALLENGE_COMPLETED_CLASS;
          break;
        default:
          newStatus = CHALLENGE_AVAILABLE_CLASS;
      }
    } else if (found && found.isFirst) {
      newStatus = CHALLENGE_AVAILABLE_CLASS;
    }

    if (newStatus !== status) {
      this.setState({
        status: newStatus,
        tooltipStatus: newStatus
      });
    }
  }

  render() {
    const {
      coordinates,
      familyId,
      type
    } = this.props;
    const { status, tooltipStatus } = this.state;

    const styles = {
      left: coordinates[0],
      top: coordinates[1]
    };

    return (
      <Tooltip fieldName={`generic.tooltips.productnetwork.${tooltipStatus}`}>
        <span
          style={styles}
          className={`challengeIcon ${familyId} ${status} ${type}`}
        />
      </Tooltip>
    );
  }
}

ChallengeIcon.propTypes = {
  challenges: PropTypes.array.isRequired,
  challengeId: PropTypes.string.isRequired,
  coordinates: PropTypes.array.isRequired,
  familyId: PropTypes.string.isRequired,
  lastChallengesModified: PropTypes.array.isRequired,
  mapInitialised: PropTypes.bool.isRequired,
  type: PropTypes.string
};

ChallengeIcon.defaultProps = {
  type: ''
};

export const mapStateToProps = state => ({
  challenges: state.mapData.challenges,
  lastChallengesModified: state.mapData.lastChallengesModified,
  mapInitialised: state.mapData.initialised
});

export default connect(
  mapStateToProps
)(ChallengeIcon);
