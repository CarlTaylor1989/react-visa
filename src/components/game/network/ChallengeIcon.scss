@import '../../../styles/buttons';

$star-icon: '../../../styles/icons/star.svg';
$star-icon-height: 12px;
$star-icon-width: 12px;
$star-icon-height-large: 32px;
$star-icon-width-large: 32px;
$star-icon-height-large: 12px;
$star-icon-width-large: 12px;
$challenge-icon-height: 10px;
$challenge-icon-width: $challenge-icon-height;
$challenge-icon-height-perfect: 14px;
$challenge-icon-width-perfect: $challenge-icon-height;
$challenge-icon-height-large: 30px;
$challenge-icon-width-large: $challenge-icon-height-large;
$challenge-icon-height-large-perfect: 40px;
$challenge-icon-width-large-perfect: $challenge-icon-height-large-perfect;

.challengeIcon {
  @extend %button-reset;
  @include border-radius($challenge-icon-height-perfect / 2);
  box-sizing: border-box;
  height: $challenge-icon-height;
  position: absolute;
  width: $challenge-icon-width;
  z-index: 3;

  &.m1f1 {
    @include challenge-icon($colour-family-1);
  }

  &.m1f2 {
    @include challenge-icon($colour-family-2);
  }

  &.m1f3 {
    @include challenge-icon($colour-family-3);
  }

  &.m1f4 {
    @include challenge-icon($colour-family-4);
  }

  &.m1f5 {
    @include challenge-icon($colour-family-5);
  }

  &.available {
    border: 1px solid $colour-white;
  }

  &.completed {
    background-color: $colour-white;
    border: 0;
  }

  &.perfect {
    border: 0;
    height: $challenge-icon-height-perfect;
    margin-left: -($challenge-icon-width-perfect - $challenge-icon-width) / 2;
    margin-top: -($challenge-icon-height-perfect - $challenge-icon-height) / 2;
    width: $challenge-icon-height-perfect;

    &::after {
      background-image: url($star-icon);
      background-position: center;
      background-repeat: no-repeat;
      background-size: $star-icon-width $star-icon-height;
      content: '';
      display: block;
      height: $challenge-icon-height-perfect;
      left: 0;
      position: absolute;
      top: 0;
      width: $challenge-icon-height-perfect;
    }
  }

  &.large {
    @include border-radius($challenge-icon-height-large-perfect / 2);
    height: $challenge-icon-height-large;
    width: $challenge-icon-width-large;

    &.available {
      border-width: 2px;
    }

    &.perfect {
      height: $challenge-icon-height-large-perfect;
      margin-left: -($challenge-icon-width-large-perfect - $challenge-icon-width-large) / 2;
      margin-top: -($challenge-icon-height-large-perfect - $challenge-icon-height-large) / 2;
      width: $challenge-icon-width-large-perfect;

      &::after {
        background-size: $star-icon-width-large $star-icon-height-large;
        height: $challenge-icon-height-large-perfect;
        width: $challenge-icon-width-large-perfect;
      }
    }
  }
}
