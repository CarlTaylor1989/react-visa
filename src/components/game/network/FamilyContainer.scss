@import '../../../styles/mixins';

$m1f1-family-width: 678px;
$m1f1-family-height: 375px;
$m1f2-family-width: 604px;
$m1f2-family-height: 265px;
$m1f3-family-width: 310px;
$m1f3-family-height: 350px;
$m1f4-family-width: 536px;
$m1f4-family-height: 369px;
$m1f5-family-width: 639px;
$m1f5-family-height: 451px;

.familyContainer {
  @include network-map;
  position: absolute;

  .family > svg {
    pointer-events: none;
  }

  .productName {
    font-size: 14px;
    font-weight: 600;

    &.center {
      text-align: center;
    }

    &.left {
      text-align: left;
    }
  }

  svg {
    [id$='_glow'],
    [id$='_thick'],
    [id$='_thin'] {
      @include opacity(0);
    }
  }

  &.m1f1 {
    height: $m1f1-family-height;
    left: 558px;
    top: -15px;
    width: $m1f1-family-width;
  }

  &.m1f2 {
    height: $m1f2-family-height;
    left: 596px;
    top: 278px;
    width: $m1f2-family-width;
  }

  &.m1f3 {
    height: $m1f3-family-height;
    left: 499px;
    top: 358px;
    width: $m1f3-family-width;
  }

  &.m1f4 {
    height: $m1f4-family-height;
    left: 84px;
    top: 310px;
    width: $m1f4-family-width;
  }

  &.m1f5 {
    height: $m1f5-family-height;
    left: 0;
    top: 27px;
    width: $m1f5-family-width;
  }
}
