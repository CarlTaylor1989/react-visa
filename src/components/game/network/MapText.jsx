import React from 'react';
import PropTypes from 'prop-types';

import './MapText.scss';

const MapText = (props) => {
  const { nameCoord, type } = props;
  const styles = {
    left: nameCoord.coordinates[0],
    top: nameCoord.coordinates[1]
  };
  return (
    <span
      dangerouslySetInnerHTML={{ __html: nameCoord.name }} // eslint-disable-line react/no-danger
      style={styles}
      className={`mapText ${type}`}
    />
  );
};

MapText.propTypes = {
  nameCoord: PropTypes.object.isRequired,
  type: PropTypes.string.isRequired
};

export default MapText;
