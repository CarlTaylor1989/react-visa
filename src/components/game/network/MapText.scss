@import '../../../styles/variables';

.mapText {
  &.familyName,
  &.productName {
    color: $colour-research;
    position: absolute;
    text-align: left;
    z-index: 1;
  }

  &.productName {
    color: $colour-white;
    font-size: 14px;
    font-weight: 600;
    line-height: 16px;
    max-width: 160px;

    &:not(.large) {
      &.small {
        max-width: 120px;
      }
    }

    &.smaller {
      max-width: 60px;
    }

    &.large {
      font-size: 36px;
      line-height: 40px;
      max-width: 400px;
      text-align: center;
      width: 400px;

      &.medium {
        width: 300px;
      }

      &.small {
        width: 250px;
      }

      &.smaller {
        width: 170px;
      }

      &.left {
        text-align: left;
      }
    }
  }

  &.familyName {
    font-size: 16px;
    font-weight: 800;

    &:not(.small) {
      white-space: nowrap;
    }

    &.small {
      width: 250px;
    }

    &.type1 {
      color: $colour-family-1;
    }

    &.type2 {
      color: $colour-family-2;
    }

    &.type3 {
      color: $colour-family-3;
    }

    &.type4 {
      color: $colour-family-4;
    }

    &.type5 {
      color: $colour-family-5;
    }
  }
}
