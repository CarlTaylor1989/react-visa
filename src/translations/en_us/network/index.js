import * as network from './network.json';
import * as productcompletion from './productcompletion.json';
import * as introsequence from './introsequence.json';
import productpages from './productpages/index';

export default {
  ...network,
  ...productcompletion,
  productpages,
  introsequence
};
