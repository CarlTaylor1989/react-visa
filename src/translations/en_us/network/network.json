{
  "productNetworkBtn": "MY PRODUCT NETWORK",

  "perfectChallenges": "PERFECT CHALLENGES",
  "perfectChallengesDivider": " / ",

  "instructions": {
    "unlockFirst": "EXPLORE THE RAIL TO UNLOCK YOUR FIRST PRODUCT..."
  },
  "start": "START EXPLORING",
  "backBtn": "BACK",
  "nextBtn": "NEXT",
  "skipBtn": "TAKE ME STRAIGHT TO THE GAME",
  "introSequenceTitle": "Welcome to the interactive Network Map."
}
