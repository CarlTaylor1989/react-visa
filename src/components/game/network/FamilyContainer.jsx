import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import InlineSVG from 'svg-inline-react';
import MapText from './MapText';
import ProductIcon from './ProductIcon';
import ChallengeIcon from './ChallengeIcon';
import challengeData from '../../../translations/en_us/network/maps/map1/challenges/index';
import productData from '../../../translations/en_us/network/maps/map1/products/index';
import familyData from '../../../translations/en_us/network/maps/map1/families.json';
import {
  CHALLENGE_PERFECT_STATUS,
  CHALLENGE_COMPLETION_STATUS
} from '../../../config/constants';

import './FamilyContainer.scss';

export class FamilyContainer extends Component {
  constructor(props) {
    super(props);
    this.state = {
      completedChallenges: []
    };
  }

  componentDidMount() {
    this.setState({
      completedChallenges: this.getCompletedChallenges()
    });
  }

  getCompletedChallenges() {
    const { challenges, familyId } = this.props;

    const completedChallenges = challenges.filter(ch => ch.familyId === familyId
      && (ch.status === CHALLENGE_COMPLETION_STATUS || ch.status === CHALLENGE_PERFECT_STATUS));
    return completedChallenges.map(ch => ch.id);
  }

  render() {
    const { familyId, mapId } = this.props;
    const { completedChallenges } = this.state;
    const currentFamilyData = familyData[familyId];
    // Create family name component
    const familyName = (
      <MapText
        key={currentFamilyData.id}
        nameCoord={currentFamilyData.nameCoord}
        type={`familyName ${currentFamilyData.type} ${currentFamilyData.nameCoord.classes
          ? currentFamilyData.nameCoord.classes.join(' ')
          : ''}`}
      />
    );

    // Create product icon and product name components
    const productIconsNames = [];
    productData[familyId].forEach((data) => {
      productIconsNames.push((
        <ProductIcon
          key={`${data.familyId}${data.id}`}
          coordinates={data.coordinates}
          productId={data.id}
          familyId={data.familyId}
          status={data.status || 0}
          mapId={mapId}
          priorityPosition={data.priorityPosition}
          useButton
        />
      ), (
        <MapText
          key={`name${data.familyId}${data.id}`}
          nameCoord={data.nameCoord}
          type={`productName ${data.nameCoord.classes
            ? data.nameCoord.classes.join(' ')
            : ''} ${data.nameCoord.type ? data.nameCoord.type : ''}`}
        />
      ));
    });

    // Create challenge icon component
    const challengeIcons = challengeData[familyId].map(data => (
      <ChallengeIcon
        key={`${familyId}${data.id}`}
        challengeId={data.id}
        coordinates={data.coordinates}
        familyId={familyId}
      />
    ));
    const svg = require(`../../../resources/${mapId}/${familyId}/${familyId}.svg`); // eslint-disable-line

    return (
      <div className={`familyContainer ${familyId} ${completedChallenges.join(' ')}`}>
        <InlineSVG className="family" src={svg} />
        {challengeIcons}
        {productIconsNames}
        {familyName}
      </div>
    );
  }
}

FamilyContainer.propTypes = {
  challenges: PropTypes.array.isRequired,
  familyId: PropTypes.string.isRequired,
  mapId: PropTypes.string.isRequired
};

export const mapStateToProps = state => ({
  challenges: state.mapData.challenges
});

export default connect(
  mapStateToProps
)(FamilyContainer);
