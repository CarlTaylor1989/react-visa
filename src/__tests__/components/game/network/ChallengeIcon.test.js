/* global toJson */
import React from 'react'; // eslint-disable-line no-unused-vars
import { ChallengeIcon, mapStateToProps } from '../../../../components/game/network/ChallengeIcon';
import {
  CHALLENGE_PERFECT_STATUS,
  CHALLENGE_COMPLETION_STATUS,
  CHALLENGE_AVAILABLE_STATUS,
  CHALLENGE_UNAVAILABLE_STATUS
} from '../../../../config/constants';


let mockProps;
beforeEach(() => {
  mockProps = {
    challenges: [],
    challengeId: 'c1',
    coordinates: [100, 100],
    familyId: 'm1f1',
    lastChallengesModified: [],
    currentProduct: '',
    mapInitialised: false
  };
});

test('renders the ChallengeIcon component with the available state', () => {
  const wrapper = shallow(<ChallengeIcon {...mockProps} />);
  expect(toJson(wrapper, { noKey: false })).toMatchSnapshot();
});

test('method componentDidUpdate()', () => {
  const wrapper = shallow(<ChallengeIcon {...mockProps} />);
  wrapper.instance().updateChallengeState = jest.fn();
  wrapper.setProps({
    lastChallengesModified: ['c1', 'c2']
  });
  wrapper.update();
  expect(wrapper.instance().updateChallengeState).toHaveBeenCalled();
});

describe('method updateChallengeState()', () => {
  test('sets the status to completed', () => {
    const challenges = [{
      id: 'c1',
      familyId: 'm1f1',
      status: CHALLENGE_COMPLETION_STATUS
    }];
    const wrapper = shallow(<ChallengeIcon {...mockProps} challenges={challenges} />);
    wrapper.instance().updateChallengeState();
    expect(toJson(wrapper, { noKey: false })).toMatchSnapshot();
    expect(wrapper.instance().state.status).toBe('completed');
  });

  test('sets the status to perfect', () => {
    const challenges = [{
      id: 'c1',
      familyId: 'm1f1',
      status: CHALLENGE_PERFECT_STATUS
    }];
    const wrapper = shallow(<ChallengeIcon {...mockProps} challenges={challenges} />);
    wrapper.instance().updateChallengeState();
    expect(toJson(wrapper, { noKey: false })).toMatchSnapshot();
    expect(wrapper.instance().state.status).toBe('perfect');
  });

  test('sets the status to available', () => {
    const challenges = [{
      id: 'c1',
      familyId: 'm1f1',
      status: CHALLENGE_AVAILABLE_STATUS
    }];
    const wrapper = shallow(<ChallengeIcon {...mockProps} challenges={challenges} />);
    wrapper.instance().updateChallengeState();
    expect(toJson(wrapper, { noKey: false })).toMatchSnapshot();
    expect(wrapper.instance().state.status).toBe('available');
  });

  test('sets the status to available for the first tree node', () => {
    const challenges = [{
      id: 'c1',
      familyId: 'm1f1',
      status: CHALLENGE_UNAVAILABLE_STATUS,
      isFirst: true
    }];
    const wrapper = shallow(<ChallengeIcon {...mockProps} challenges={challenges} />);
    wrapper.instance().updateChallengeState();
    expect(toJson(wrapper, { noKey: false })).toMatchSnapshot();
    expect(wrapper.instance().state.status).toBe('available');
  });
});

test('should mapStateToProps', () => {
  const state = {
    mapData: {
      challenges: [],
      initialised: false,
      lastChallengesModified: []
    }
  };
  const props = mapStateToProps(state);
  expect(props).toEqual({
    challenges: [],
    mapInitialised: false,
    lastChallengesModified: []
  });
});
