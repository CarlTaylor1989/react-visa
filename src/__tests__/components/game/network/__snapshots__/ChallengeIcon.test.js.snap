// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`method updateChallengeState() sets the status to available 1`] = `
<Connect(Tooltip)
  fieldName="generic.tooltips.productnetwork.available"
>
  <span
    className="challengeIcon m1f1 available "
    style={
      Object {
        "left": 100,
        "top": 100,
      }
    }
  />
</Connect(Tooltip)>
`;

exports[`method updateChallengeState() sets the status to available for the first tree node 1`] = `
<Connect(Tooltip)
  fieldName="generic.tooltips.productnetwork.available"
>
  <span
    className="challengeIcon m1f1 available "
    style={
      Object {
        "left": 100,
        "top": 100,
      }
    }
  />
</Connect(Tooltip)>
`;

exports[`method updateChallengeState() sets the status to completed 1`] = `
<Connect(Tooltip)
  fieldName="generic.tooltips.productnetwork.completed"
>
  <span
    className="challengeIcon m1f1 completed "
    style={
      Object {
        "left": 100,
        "top": 100,
      }
    }
  />
</Connect(Tooltip)>
`;

exports[`method updateChallengeState() sets the status to perfect 1`] = `
<Connect(Tooltip)
  fieldName="generic.tooltips.productnetwork.perfect"
>
  <span
    className="challengeIcon m1f1 perfect "
    style={
      Object {
        "left": 100,
        "top": 100,
      }
    }
  />
</Connect(Tooltip)>
`;

exports[`renders the ChallengeIcon component with the available state 1`] = `
<Connect(Tooltip)
  fieldName="generic.tooltips.productnetwork.available"
>
  <span
    className="challengeIcon m1f1  "
    style={
      Object {
        "left": 100,
        "top": 100,
      }
    }
  />
</Connect(Tooltip)>
`;
